metadata/sources.go

```diff
@@ -0,0 +1,39 @@
+package metadata
+
+type (
+	RequestBinSource struct {
+		Type SourceType `json:"type,omitempty"`
+		Bin  string     `json:"bin,omitempty"`
+	}
+
+	RequestCardSource struct {
+		Type   SourceType `json:"type,omitempty"`
+		Number string     `json:"number,omitempty"`
+	}
+
+	RequestIdSource struct {
+		Type SourceType `json:"type,omitempty"`
+		Id   string     `json:"id,omitempty"`
+	}
+
+	RequestTokenSource struct {
+		Type  SourceType `json:"type,omitempty"`
+		Token string     `json:"token,omitempty"`
+	}
```

Comment:
The idea is right, but there are a few things to change here: 
- You should declare an interface, let's say `MetadataSourceType interface{}` and all these types should implement that interface (you can declare a dummy method to implement it, like GetType() SourceType). Then in `CardMetadataRequest` the Source attribute should be of `MetadataSourceType`, so that only the types that implement these interface can be passed as Source for the request.
- These types here should be "private" (start with a lower case letter) so that merchants are forced to use the constructor methods to instantiate them.

You can check `conditions.go` in these [PR](https://github.com/checkout/checkout-sdk-go/pull/22/files#diff-895dae80c28fef139d121de4ffba41f6583caf978cd5b524e0057e5bef449209) to get an idea of how it should be done

Reply:
👍